pkg/component/ref-objects.go

```diff
@@ -0,0 +1,127 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package component
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
+)
+
+// GetLabelSelectorFromRefObjectSelector extract labelSelector from `labelSelector` first. If empty, extract from `selector`
+func GetLabelSelectorFromRefObjectSelector(selector v1alpha1.ObjectReferrer) map[string]string {
+	if selector.LabelSelector != nil {
+		return selector.LabelSelector
+	}
+	return selector.DeprecatedLabelSelector
+}
+
+// GetGroupVersionKindFromRefObjectSelector extract GroupVersionKind by Resource if provided, otherwise, extract from APIVersion and Kind directly
+func GetGroupVersionKindFromRefObjectSelector(mapper meta.RESTMapper, selector v1alpha1.ObjectReferrer) (schema.GroupVersionKind, error) {
+	if selector.Resource != "" {
+		return mapper.KindFor(schema.GroupVersionResource{Group: selector.Group, Resource: selector.Resource})
+	}
+	if selector.APIVersion != "" && selector.Kind != "" {
+		gv, err := schema.ParseGroupVersion(selector.APIVersion)
+		if err != nil {
+			return schema.GroupVersionKind{}, errors.Wrapf(err, "invalid APIVersion")
+		}
+		return gv.WithKind(selector.Kind), nil
+	}
+	return schema.GroupVersionKind{}, errors.Errorf("object kind identifier unavailable")
+}
+
+// ValidateRefObjectSelector validate if exclusive fields are set for the selector
+func ValidateRefObjectSelector(selector v1alpha1.ObjectReferrer) error {
+	labelSelector := GetLabelSelectorFromRefObjectSelector(selector)
+	if labelSelector != nil && selector.Name != "" {
+		return errors.Errorf("invalid object selector for ref-objects, name and labelSelector cannot be both set")
+	}
+	if labelSelector == nil && selector.Name == "" {
+		return errors.Errorf("invalid object selector for ref-objects, either name or labelSelector must be set")
+	}
+	return nil
+}
+
+// ClearRefObjectForDispatch reset the objects for dispatch
+func ClearRefObjectForDispatch(un *unstructured.Unstructured) {
+	un.SetResourceVersion("")
+	un.SetGeneration(0)
+	un.SetOwnerReferences(nil)
+	un.SetDeletionTimestamp(nil)
+	un.SetManagedFields(nil)
+	un.SetUID("")
+	unstructured.RemoveNestedField(un.Object, "metadata", "creationTimestamp")
+	unstructured.RemoveNestedField(un.Object, "status")
+	// TODO(somefive): make the following logic more generalizable
+	if un.GetKind() == "Service" && un.GetAPIVersion() == "v1" {
```

Comment:
consider copy/paste the following code which is proven working correct for yrs.

https://github.com/openshift/library-go/blob/d48a1fb9b7c2c89f5914cc670a4056bcccdd0a34/pkg/operator/resource/resourceapply/core.go#L130-L177

Reply:
@yue9944882 If I'm understanding it correct, .spec.ports is not copied in this implementation? Seems only .spec.type and .spec.selector are copied.